File: crates/avin_model/src/bar_direction.rs
// ───────────────────────────────────────────────────────────────────────────
// AVIN
// Understand the market before trading it.
//
// https://avin.info
// ───────────────────────────────────────────────────────────────────────────

/// Bar direction.
///
/// # ru
/// Тип бара: бычий, доджи, медвежий.
#[repr(i8)]
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum BarDirection {
    /// Бычий
    Bull = 1,
    /// Доджи (открытие == закрытие)
    Doji = 0,
    /// Медвежий
    Bear = -1,
}

impl std::fmt::Display for BarDirection {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        match self {
            BarDirection::Bull => write!(f, "Bull"),
            BarDirection::Doji => write!(f, "Doji"),
            BarDirection::Bear => write!(f, "Bear"),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn values() {
        assert_eq!(BarDirection::Bull as i8, 1);
        assert_eq!(BarDirection::Doji as i8, 0);
        assert_eq!(BarDirection::Bear as i8, -1);
    }

    #[test]
    fn display() {
        assert_eq!(BarDirection::Bull.to_string(), "Bull");
        assert_eq!(BarDirection::Doji.to_string(), "Doji");
        assert_eq!(BarDirection::Bear.to_string(), "Bear");
    }
}

File: python/avin/domain/bar_direction.py
# ────────────────────────────────────────────────────────────────────────────
#  AVIN
#  Understand the market before trading it.
#
#  https://avin.info
# ────────────────────────────────────────────────────────────────────────────

from __future__ import annotations

from enum import Enum

from avin._native import PyBarDirection


class BarDirection(Enum):
    """
    Bar direction.

    Indicates whether a bar is bullish, neutral, or bearish.

    Attributes
    ----------
    BULL
        The bar closes above its opening price.
    NEUTRAL
        The bar closes at its opening price.
    BEAR
        The bar closes below its opening price.

    Examples
    --------
    >>> direction = BarDirection.BULL
    >>> direction.name
    'BULL'
    >>> direction.value
    1
    >>> str(direction)
    'Bull'
    """

    BULL = PyBarDirection.Bull
    NEUTRAL = PyBarDirection.Neutral
    BEAR = PyBarDirection.Bear

    _inner: PyBarDirection

    def __new__(cls, inner: PyBarDirection):
        obj = object.__new__(cls)
        obj._value_ = inner.value()
        obj._inner = inner

        return obj

    def __str__(self) -> str:
        return self._inner.display()

    @classmethod
    def _from_native(cls, inner: PyBarDirection) -> BarDirection:
        for direction in cls:
            if direction._inner.eq(inner):
                return direction

        raise RuntimeError(
            "native bar direction is missing from BarDirection"
        )
